### DISCUSSION TITLE
undefined 值的作用

### DISCUSSION THREAD

**Discussion by jiacai2050** [zigcc/forum#8] [观点交流]
- https://www.reddit.com/r/Zig/comments/w8agp0/when_would_you_use_an_undefined_value/

> undefined also communicates intent, allowing safety checks & tooling to help you catch bugs. if you use valgrind with zig for example, it will catch when you accidentally branch on an undefined value. however it would not catch when you branch on a zero value as above, even when such zero values are incorrect. furthermore, zig makes undefined in debug mode be 0xaa bytes which tends to catch bugs early on, vs zero initializing things tends to create bugs that are harder to discover.

- https://research.swtch.com/sparse
  - https://github.com/Srekel/zig-sparse-set

**Comment by hhstore** [comment]
- https://github.com/better-zig/ziglings-solutions/blob/solutions/exercises/050_no_value.zig#L55
- `ziglings` 第50个练习题, 注释有 `undefined 以及 null, void` 区别的说明. 
- 不过, 个人感觉 `undefined` 不应该存在, 有点倒退. 
- 是为了更好的`兼容 C` 吗? Zig 很多设计, 跟 C 保持高度一致.

> 应用示例:(链表定义) 

- https://github.com/better-zig/ziglings-solutions/blob/solutions/exercises/044_quiz5.zig#L16
- https://github.com/better-zig/ziglings-solutions/blob/solutions/exercises/046_optionals2.zig#L28
- 前者是使用 `undefined`, 后者是使用 `optional ` + `null` 替代. 
- so, 后者是更好写法. 后者也类似 dart 的 `null safety` 设计.

**Reply by jiacai2050** [reply]
功能上 undefined 是可以等价 option+null 的，undefined 我理解主要还是性能上的需要，用了 option 后面使用时，势必会有 if 判断，这就有开销了，undefined 则没有这个需要，用户按照业务逻辑去正确的初始化 undefined 即可。比如读取文件内容

```zig
var buf: [1024]u8 = undefined;
const bytes_read = try f.readAll(buf[0..]);
```
这里用 undefined 就要比 option + null 的方式优雅些。

**Reply by xiongxin** [reply]
undefined 就是和C未定义的变量一样吧，当一个变量为 undefined 时，然后对这个变量操作时，会有一些意想不到的结果

**Reply by htqx** [reply]
undefined 未定义。它是一个静态检测，等于告诉编译器不要让代码在没有定义前使用。
null可没有这个保证。null是一个值，真正的状态。所以还是不同的。当然两则比较接近。

如果是希望变量具备 null 值（或者有时会处于失效状态，比如指针类），就定义?T，如果只是暂时没有初始化，那么就用 undefined。

**Reply by ClarkThan** [reply]
undefined 就是明确标识某个变量现在还没有初始化, 后面某个地方在使用前才会初始化,  其实就是变量申明 (对比 golang 中 var a int 声明时 默认会将 a 初始为 int 零值,  可以直接使用.  zig 严格要求必选显式初始化才能使用, 否则编译检查会报错);
null 是一个确定类型的值, 本质上就是 tag union 类型的值, (对比 Rust 的 enum Option<T>);
